File: JAVA/src/arrayMedium/nextPermutation.java
package arrayMedium;

import java.util.Arrays;

public class nextPermutation {
    public static int[] function(int[] arr){
        int pivot = -1;
        int end = arr.length-1;
        for (int i = arr.length-1; i > 0; i--) {
            if (arr[i-1] < arr[i]){
                pivot = i-1;
                break;
            }
        }
        if(pivot == -1){
            int start = 0;
            while (start < end ){
                int temp = arr[start];
                arr[start] = arr[end];
                arr[end] = temp;
                start++;
                end--;

            }
        } else {
            for (int i = arr.length - 1; i > 0; i--) {
                if( arr[i] > arr[pivot]){
                    int temp = arr[i];
                    arr[i] = arr[pivot];
                    arr[pivot] = temp;
                    break;
                }
            }

            while (pivot < end ){
                int temp = arr[pivot+1];
                arr[pivot+1] = arr[end];
                arr[end] = temp;
                pivot++;
                end--;

            }
        }



        return arr;
    }
    public static void main(String[] args) {
        int[] arr = {1,2,3};
        int[] result = function(arr);
        System.out.println(Arrays.toString(result));
    }

}

File: PYTHON/01/arrays_medium_problems/next_permutation.py
class Solution:
    def function(self,arr):
        end = len(arr)-1
        pivot = -1

        for i in range(end,0,-1):
            if arr[i] > arr[i-1]:
                pivot = i-1
                break

        if pivot == -1:
            start = 0
            while start < end:
                arr[start], arr[end] = arr[end], arr[start]

                start+=1
                end-=1
        else:
            for i in range(end,0,-1):
                if arr[i] > arr[pivot]:
                    arr[i], arr[pivot] = arr[pivot], arr[i]
                    break
            while pivot < end:
                arr[pivot+1], arr[end] = arr[end], arr[pivot+1]
                pivot+=1
                end-=1
        return arr

if __name__ == '__main__':
    arr = [3,2,1]
    obj = Solution()
    result = obj.function(arr)
    print(result)